experimental/dds/tree2/src/feature-libraries/object-forest/objectForest.ts

```diff
@@ -141,41 +127,74 @@ class ObjectForest extends SimpleDependee implements IEditableForest {
+			},
+			attachEdit(source: DetachedRangeUpPath, destination: PlaceIndex): void {
 				const [parent, key] = cursor.getParent();
-				const sourceField = getMapTreeField(parent, key, false);
-				const startIndex = index;
-				const endIndex = index + count;
-				assertValidIndex(startIndex, sourceField, true);
-				assertValidIndex(endIndex, sourceField, true);
-				assert(
-					startIndex <= endIndex,
-					0x371 /* detached range's end must be after its start */,
-				);
-				const newField = sourceField.splice(startIndex, endIndex - startIndex);
-				const field = this.forest.addFieldAsDetached(newField);
-				if (id !== undefined) {
-					moves.set(id, field);
-				} else {
-					this.forest.delete(field);
-				}
+				const currentField = getMapTreeField(parent, key, true);
+				assertValidIndex(destination, currentField, true);
+				const sourceField = getMapTreeField(this.forest.roots, source.field, false);
+				assertValidIndex(source.start, sourceField, true);
+				assertValidIndex(source.end, sourceField, true);
+				assert(source.end >= source.start, "Range end must be >= its start");
+				const content = sourceField.splice(source.start, source.end - source.start);
 				if (sourceField.length === 0) {
+					this.forest.roots.fields.delete(source.field);
+				}
+				// TODO: this will fail for very large insertions due to argument limits.
+				currentField.splice(destination, 0, ...content);
+			},
+			detachEdit(source: Range, destination: DetachedPlaceUpPath | undefined): void {
+				const [parent, key] = cursor.getParent();
+				const currentField = getMapTreeField(parent, key, true);
+				assertValidIndex(source.start, currentField, true);
+				assertValidIndex(source.end, currentField, true);
+				assert(source.end >= source.start, "Range end must be >= its start");
+				const content = currentField.splice(source.start, source.end - source.start);
+				if (currentField.length === 0) {
 					parent.fields.delete(key);
 				}
+				if (destination !== undefined) {
+					const destinationField = getMapTreeField(
+						this.forest.roots,
+						destination.field,
+						true,
+					);
+					assertValidIndex(destination.index, destinationField, true);
+					destinationField.splice(
+						destination.index,
+						0,
+						// TODO: this will fail for very large insertions due to argument limits.
+						...content,
+					);
+				}
 			},
-			onMoveIn(index: number, count: number, id: Delta.MoveId): void {
-				const toAttach = moves.get(id) ?? fail("move in without move out");
-				moves.delete(id);
-				const countMoved = moveIn(index, toAttach, this.cursor);
-				assert(countMoved === count, 0x369 /* counts must match */);
+			replace(
+				newContentSource: DetachedRangeUpPath,
+				oldContentRange: Range,
+				oldContentDestination: DetachedPlaceUpPath,
+			): void {
+				this.forest.invalidateDependents();
+				this.detachEdit(oldContentRange, oldContentDestination);
+				this.attachEdit(newContentSource, oldContentRange.start);
```

Comment:
since detachEdit takes in a destination thats a place in a detached field, presumably it can move stuff to a non-empty field.
This seems odd, but maybe this is part of the one node at a time temporary thing?

Anyway, if the destination field is the same as the source the index used for the destination on this line can be wrong.

I think this warrants an assert that these are not the same as a justification for holding onto that oldContent range across an edit.